Author: Eric Pruitt, https://github.com/ericpruitt/
Description: Adds an additional property named "priority" to window rule
definitions. When a new window is managed by dwm, its location in window list
is determined by its priority: the higher the priority, the earlier in the list
it is added. A default priority must be specified in config.h like so:

    static const int defaultpriority = 50;

When using the "zoom" function, a window cannot be zoomed above a window with a
higher priority.

diff --git a/dwm.c b/dwm.c
index 0362114..f0077c5 100644
--- a/dwm.c
+++ b/dwm.c
@@ -87,7 +87,7 @@ typedef struct Client Client;
 struct Client {
 	char name[256];
 	float mina, maxa;
-	int x, y, w, h;
+	int x, y, w, h, priority;
 	int oldx, oldy, oldw, oldh;
 	int basew, baseh, incw, inch, maxw, maxh, minw, minh;
 	int bw, oldbw;
@@ -139,6 +139,7 @@ typedef struct {
 	unsigned int tags;
 	int isfloating;
 	int monitor;
+	int priority;
 } Rule;
 
 /* function declarations */
@@ -147,6 +148,7 @@ static int applysizehints(Client *c, int *x, int *y, int *w, int *h, int interac
 static void arrange(Monitor *m);
 static void arrangemon(Monitor *m);
 static void attach(Client *c);
+static void attachbypriority(Client *c, int attachafterifequal);
 static void attachstack(Client *c);
 static void buttonpress(XEvent *e);
 static void checkotherwm(void);
@@ -297,6 +299,7 @@ applyrules(Client *c)
 		&& (!r->instance || strstr(instance, r->instance)))
 		{
 			c->isfloating = r->isfloating;
+			c->priority = r->priority ? r->priority : c->priority;
 			c->tags |= r->tags;
 			for (m = mons; m && m->num != r->monitor; m = m->next);
 			if (m)
@@ -406,6 +409,39 @@ attach(Client *c)
 }
 
 void
+attachbypriority(Client *c, int attachafterifequal) {
+	Client *k, *p = NULL;
+	if (c->isfloating || c->isfullscreen) {
+		attach(c);
+		return;
+	}
+
+	for (k = c->mon->clients; k; k = k->next) {
+		if (k->isfloating || k->isfullscreen || k == c) {
+			continue;
+		}
+		if ((c->priority > k->priority) ||
+		   ((c->priority == k->priority) && !attachafterifequal)) {
+			if (!p) {
+				attach(c);
+			} else {
+				c->next = p->next;
+				p->next = c;
+			}
+			return;
+		}
+		p = k;
+	}
+
+	if (p) {
+		c->next = NULL;
+		p->next = c;
+	} else {
+		attach(c);
+	}
+}
+
+void
 attachstack(Client *c)
 {
 	c->snext = c->mon->stack;
@@ -1037,6 +1073,7 @@ manage(Window w, XWindowAttributes *wa)
 	XWindowChanges wc;
 
 	c = ecalloc(1, sizeof(Client));
+	c->priority = defaultpriority;
 	c->win = w;
 	updatetitle(c);
 	if (XGetTransientForHint(dpy, w, &trans) && (t = wintoclient(trans))) {
@@ -1076,7 +1113,7 @@ manage(Window w, XWindowAttributes *wa)
 		c->isfloating = c->oldstate = trans != None || c->isfixed;
 	if (c->isfloating)
 		XRaiseWindow(dpy, c->win);
-	attach(c);
+	attachbypriority(c, 1);
 	attachstack(c);
 	XChangeProperty(dpy, root, netatom[NetClientList], XA_WINDOW, 32, PropModeAppend,
 	                (unsigned char *) &(c->win), 1);
@@ -1220,7 +1257,7 @@ void
 pop(Client *c)
 {
 	detach(c);
-	attach(c);
+	attachbypriority(c, 0);
 	focus(c);
 	arrange(c->mon);
 }
